src/libraries/System.Linq/src/System/Linq/AggregateBy.cs

```diff
@@ -0,0 +1,78 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+
+namespace System.Linq
+{
+    public static partial class Enumerable
+    {
+        public static IEnumerable<KeyValuePair<TKey, TAccumulate>> AggregateBy<TSource, TKey, TAccumulate>(
+            this IEnumerable<TSource> source,
+            Func<TSource, TKey> keySelector,
+            TAccumulate seed,
+            Func<TAccumulate, TSource, TAccumulate> func,
+            IEqualityComparer<TKey>? keyComparer = null) where TKey : notnull
+            => AggregateBy(source, keySelector, _ => seed, func, keyComparer);
+
+        public static IEnumerable<KeyValuePair<TKey, TAccumulate>> AggregateBy<TSource, TKey, TAccumulate>(
+            this IEnumerable<TSource> source,
+            Func<TSource, TKey> keySelector,
+            Func<TKey, TAccumulate> seedSelector,
+            Func<TAccumulate, TSource, TAccumulate> func,
+            IEqualityComparer<TKey>? keyComparer = null) where TKey : notnull
+        {
+            if (source is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
+            }
+            if (keySelector is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.keySelector);
+            }
+            if (seedSelector is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.seedSelector);
+            }
+            if (func is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.func);
+            }
+
+            return AggregateByIterator(source, keySelector, seedSelector, func, keyComparer);
+        }
+
+        private static IEnumerable<KeyValuePair<TKey, TAccumulate>> AggregateByIterator<TSource, TKey, TAccumulate>(IEnumerable<TSource> source, Func<TSource, TKey> keySelector, Func<TKey, TAccumulate> seedSelector, Func<TAccumulate, TSource, TAccumulate> func, IEqualityComparer<TKey>? keyComparer) where TKey : notnull
+        {
+            using IEnumerator<TSource> enumerator = source.GetEnumerator();
+
+            if (!enumerator.MoveNext())
+            {
+                yield break;
+            }
+
+            foreach (KeyValuePair<TKey, TAccumulate> countBy in PopulateDictionary<TSource, TKey, TAccumulate>(enumerator, keySelector, seedSelector, func, keyComparer))
+            {
+                yield return countBy;
+            }
+        }
+
+        private static Dictionary<TKey, TAccumulate> PopulateDictionary<TSource, TKey, TAccumulate>(IEnumerator<TSource> enumerator, Func<TSource, TKey> keySelector, Func<TKey, TAccumulate> seedSelector, Func<TAccumulate, TSource, TAccumulate> func, IEqualityComparer<TKey>? keyComparer) where TKey : notnull
```

Comment:
I would recommend making this method a local static, or at least use a name that points to this method being related to `AggregateBy`.

Reply:
Moved to a local static method.